backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoicecandidate/async/spi/impl/RecreateInvoiceWorkpackageProcessor.java

```diff
@@ -0,0 +1,138 @@
+import de.metas.invoicecandidate.model.I_C_Invoice_Candidate;
+import de.metas.lock.api.ILockAutoCloseable;
+import de.metas.lock.api.ILockCommand;
+import de.metas.lock.api.ILockManager;
+import de.metas.lock.api.LockOwner;
+import de.metas.logging.LogManager;
+import de.metas.process.PInstanceId;
+import de.metas.util.Check;
+import de.metas.util.Loggables;
+import de.metas.util.Services;
+import lombok.NonNull;
+import org.adempiere.ad.dao.IQueryBL;
+import org.adempiere.model.InterfaceWrapperHelper;
+import org.compiere.model.I_AD_PInstance;
+import org.compiere.model.I_C_Invoice;
+import org.compiere.model.I_C_Payment;
+import org.slf4j.Logger;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+
+import static org.compiere.util.Env.getCtx;
+
+public class RecreateInvoiceWorkpackageProcessor extends WorkpackageProcessorAdapter
+{
+	private static final Logger logger = LogManager.getLogger(RecreateInvoiceWorkpackageProcessor.class);
+
+	private final IQueryBL queryBL = Services.get(IQueryBL.class);
+	private final IAsyncBatchBL asyncBatchBL = Services.get(IAsyncBatchBL.class);
+	private final IAllocationDAO allocationDAO = Services.get(IAllocationDAO.class);
+	private final IInvoiceCandBL invoiceCandBL = Services.get(IInvoiceCandBL.class);
+	private final ILockManager lockManager = Services.get(ILockManager.class);
+
+	@Override
+	public Result processWorkPackage(@NonNull final I_C_Queue_WorkPackage workpackage, @NonNull final String trxName)
+	{
+		final int pinstanceInt = getParameters().getParameterAsInt(I_AD_PInstance.COLUMNNAME_AD_PInstance_ID, workpackage.getAD_PInstance_ID());
+		Check.assume(pinstanceInt > 0, "pinstanceInt>=0 on workpackage={}", workpackage);
+
+		final PInstanceId pinstanceId = PInstanceId.ofRepoId(pinstanceInt);
+
+		Check.assume(workpackage.getC_Async_Batch_ID() > 0, "workpackage.getC_Async_Batch_ID() > 0 for workpackage=", workpackage);
+		final AsyncBatchId asyncBatchId = AsyncBatchId.ofRepoId(workpackage.getC_Async_Batch_ID());
+
+		final ILockCommand elementsLocker = lockManager.lock()
+				.setOwner(LockOwner.newOwner(getClass().getSimpleName()))
+				.setAutoCleanup(true)
+				.setFailIfAlreadyLocked(true)
+				.setRecordsBySelection(I_C_Invoice_Candidate.class, pinstanceId);
+
+		try (final ILockAutoCloseable lock = elementsLocker.acquire().asAutoCloseable())
+		{
+			processWorkPackage0(pinstanceId, asyncBatchId);
+		}
+
+		return Result.SUCCESS;
+	}
+
+	private void processWorkPackage0(@NonNull final PInstanceId pinstanceId, @NonNull final AsyncBatchId asyncBatchId)
+	{
+		final Iterator<I_C_Invoice> invoiceIterator = retrieveSelection(pinstanceId);
+
+		while (invoiceIterator.hasNext())
+		{
+			final I_C_Invoice invoice = invoiceIterator.next();
+
+			if (!canVoidPaidInvoice(invoice))
+			{
+				Loggables.withLogger(logger, Level.DEBUG).addLog("Skipping invoice. Cannot void invoice, invoiceId=" + InvoiceId.ofRepoId(invoice.getC_Invoice_ID()));
```

Comment:
```suggestion
				Loggables.withLogger(logger, Level.DEBUG).addLog("C_Invoice_ID={}: Skipping paid invoice because we can't void it.", invoice.getC_Invoice_ID());
```